rllib/models/specs/checker.py

```diff
@@ -0,0 +1,290 @@
+
+            output_data = func(self, input_data_, **kwargs)
+
+            if cache and func.__name__ not in self.__checked_input_specs_cache__:
+                self.__checked_input_specs_cache__[func.__name__] = True
+
+            return output_data
+
+        wrapper.__checked_input_specs__ = True
+        return wrapper
+
+    return decorator
+
+
+@DeveloperAPI(stability="alpha")
+def check_output_specs(
+    output_spec: str,
+    *,
+    cache: bool = False,
+):
+    """A general-purpose spec checker decorator for Neural Network base classes.
+
+    This is a stateful decorator
+    (https://realpython.com/primer-on-python-decorators/#stateful-decorators) to
+    enforce output specs for any instance method that outputs a single dict-like object.
+
+    It also allows you to cache the validation to make sure the spec is only validated
+    once in the entire lifetime of the instance.
+
+    Examples (See more exmaples in ../tests/test_specs_dict.py):
+
+        >>> class MyModel(nn.Module):
+        ...     @property
+        ...     def output_spec(self):
+        ...         return {"obs": TensorSpec("b, d", d=64)}
+        ...
+        ...     @check_output_specs("output_spec")
+        ...     def forward(self, input_data, return_loss=False):
+        ...         return {"obs": torch.randn(32, 64)}
+
+    Args:
+        func: The instance method to decorate. It should be a callable that takes
+            `self` as the first argument, `input_data` as the second argument and any
+            other keyword argument thereafter. It should return a single dict-like
+            object (i.e. not a tuple).
+        input_spec: `self` should have an instance attribute whose name matches the
+            string in input_spec and returns the `SpecDict`, `Spec`, or simply the
+            `Type` that the `input_data` should comply with. It can alos be None or
+            empty list / dict to enforce no input spec.
+        cache: If True, only checks the data validation for the first time the
+            instance method is called.
+
+    Returns:
+        A wrapped instance method. In case of `cache=True`, after the first invokation
+        of the decorated method, the intance will have `__checked_output_specs_cache__`
+        attribute that stores which method has been invoked at least once. This is a
+        special attribute that can be used for the cache itself. The wrapped class
+        method also has a special attribute `__checked_output_specs__` that marks the
+        method as decorated.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, input_data, **kwargs):
+            if cache and not hasattr(self, "__checked_output_specs_cache__"):
+                self.__checked_output_specs_cache__ = {}
+
+            output_data = func(self, input_data, **kwargs)
+
+            if output_spec:
+                output_spec_ = getattr(self, output_spec)
```

Comment:
(Same as above here, could this be an issue when trying to check specs that don't exist?)

Reply:
fixed.